### ISSUE TITLE
Running the code

### ISSUE THREAD

**Issue by tvaranka** [Justin900429/mimicking-annotation-micro-expression-recognition#1]
Thanks a lot providing such well documented and high quality code.

However, I have an issue with the csv files. How do I get these? The original files provided by the datasets are xlsx files. Can I transfer these?
Edit: Found the files, they were right there 😆 

Also a question regarding the learning based motion magnification. Did you experiment with using Eulerian motion magnification and learning-based motion magnification? From my experiments I have noticed that the LBMM is worse, perhaps the parameters need to be fine-tuned?

**Comment by Justin900429** [comment]
Hi, did you mean the performance of MER?
Using Eulerian motion magnification should also be possible, but we use LBMM for its better stability.
The image should be cropped before magnification and training.
We had provided the cropping code for SAMM. You can refer to OpenFace for face cropping, which we used for CASME II.

**Comment by tvaranka** [comment]
I was able to crop the images for samm, but I am having issues with running the LBMM.

I am getting following error: InvalidArgumentError (see above for traceback): ConcatOp : Dimensions of inputs should match: shape[0] = [1,218,314,32] vs. shape[1] = [1,218,313,32]

It seems there is an issue with resizing as the width is only slightly off. I believe I have all the correct versions, any ideas where the issue could be?

Edit: full error
Caused by op 'ynet_3frames/decoder/concat', defined at:
  File "get_samm_img.py", line 35, in <module>
    model.setup_for_inference(checkpoint, 384, 384)
  File "...\Facial-Separated-Micro-Expression-Recognition\deep_motion_mag\magnet.py", line 206, in setup_for_inference
    self._build_feed_model()
  File "...\Facial-Separated-Micro-Expression-Recognition\deep_motion_mag\magnet.py", line 193, in _build_feed_model
    False)
  File "...\Facial-Separated-Micro-Expression-Recognition\deep_motion_mag\magnet.py", line 142, in image_transformer
    return self._decoder(self.texture_b, self.out_shape_enc)
  File "...\Facial-Separated-Micro-Expression-Recognition\deep_motion_mag\magnet.py", line 116, in _decoder
    enc = tf.concat([texture_enc, shape_enc], axis=3)
  File "...\Anaconda3\envs\deep_mag\lib\site-packages\tensorflow\python\ops\array_ops.py", line 1066, in concat
    name=name)
  File "....\Anaconda3\envs\deep_mag\lib\site-packages\tensorflow\python\ops\gen_array_ops.py", line 493, in _concat_v2
    name=name)
  File "....\Anaconda3\envs\deep_mag\lib\site-packages\tensorflow\python\framework\op_def_library.py", line 767, in apply_op
    op_def=op_def)
  File "....\Anaconda3\envs\deep_mag\lib\site-packages\tensorflow\python\framework\ops.py", line 2630, in create_op
    original_op=self._default_original_op, op_def=op_def)
  File "....\Anaconda3\envs\deep_mag\lib\site-packages\tensorflow\python\framework\ops.py", line 1204, in __init__
    self._traceback = self._graph._extract_stack()  # pylint: disable=protected-access

InvalidArgumentError (see above for traceback): ConcatOp : Dimensions of inputs should match: shape[0] = [1,218,314,32] vs. shape[1] = [1,218,313,32]
         [[Node: ynet_3frames/decoder/concat = ConcatV2[N=2, T=DT_FLOAT, Tidx=DT_INT32, _device="/job:localhost/replica:0/task:0/cpu:0"](ynet_3frames/decoder/Relu, ynet_3frames/manipulator/add_1, ynet_3frames/decoder/concat/axis)]]

**Comment by tvaranka** [comment]
https://github.com/12dmodel/deep_motion_mag/issues/20

Seems that it is an issue from the LBMM.

However, If you didn't get this error it means that something must be wrong with the cropping from my side

**Comment by Justin900429** [comment]
There is a function in [utils.py](https://github.com/12dmodel/deep_motion_mag/blob/485243bd7428d08059c313321b5e6ebfd7f61991/utils.py#L76) called `imread` which loads images.
I had forced all the images to be resized to `384x384`, below is the modified version
```python
def imread(path, is_grayscale=True):
    if (is_grayscale):
        return cv2.imread(path, flags=cv2.IMREAD_GRAYSCALE).astype(np.float)
    else:
        img = cv2.imread(path, flags=cv2.IMREAD_COLOR)
        img = cv2.resize(img, (384, 384))
        return cv2.cvtColor(img, code=cv2.COLOR_BGR2RGB).astype(np.float)
```
Hope this solve your issue :)

**Comment by tvaranka** [comment]
Thanks! It helped me get further! Currently the issue is trying to load subject 006_2_4. The path it is trying to read is 006_00180.jpg, the image is however 006_0180.jpg. Seems there is some additional zero padding coming from somewhere.

**Comment by Justin900429** [comment]
Oh yeah, the zero padding is needed. This is more convenient for me to write the code :D
The file name should be in the format `xxx_xxxxx.jpg`.

**Comment by tvaranka** [comment]
I had to further hack the imread function but it works now :D

```python
def imread(path, is_grayscale=True):    
    if (is_grayscale):
        return cv2.imread(path, flags=cv2.IMREAD_GRAYSCALE).astype(np.float)
    else:
        img = cv2.imread(path, flags=cv2.IMREAD_COLOR)
        if img is None:
            path = path[:-9] + path[-8:]
            img = cv2.imread(path, flags=cv2.IMREAD_COLOR)
        img = cv2.resize(img, (384, 384))
        return cv2.cvtColor(img, code=cv2.COLOR_BGR2RGB).astype(np.float)
```

**Comment by tvaranka** [comment]
Managed to run everything now. Downloaded the weights for samm_3, but the results are not good
Unweighted LOSO accuracy: 0.6842, f1-score: 0.4729
UAR: 0.4967

Any ideas?

**Comment by Justin900429** [comment]
All the preprocess features are obtained from cropping image??

**Comment by tvaranka** [comment]
Yes, I followed all the steps in the readme.md. Only changed the resizing and zero padding

**Comment by Justin900429** [comment]
Did you check the magnified image and generated features?
(Open it like a image file can check how it looks like)

BTW, may I ask what is `path = path[:-9] + path[-8:]` for?
This doesn't change path if length of the path is > 8.

**Comment by tvaranka** [comment]
The magnified image and optical flow look reasonable. The cropping seems to be off by quite a bit. There is still a lot of background, compared to the figures you have shown in the paper. Could there be an issue with the samm_crop file?

The line `path = path[:-9] + path[-8:]`, removes the zero padding for cases where the frame number is less than 999. The path string has double zero padding for cases with <999 frames, while the images only have single zero padding.

**Comment by Justin900429** [comment]
The crop image will be saved in the new place and will not overwrite the original one.
Did you use the new path to create the magnified images and the features?

**Comment by tvaranka** [comment]
Yes I am using the new path with cropped images. The crop however is quite bad, it seems that the samm_crop doesn't work properly.

**Comment by tvaranka** [comment]
I tried with images that I have cropped using a different algorithm. I got an F1-score of 0.6558 for samm three class, it should be 0.816. Seems like the issue is with the cropping, could you double check the samm_crop function?

**Comment by tvaranka** [comment]
Any updates @Justin900429 ?

**Comment by Justin900429** [comment]
Need some time for checking the code.
Will update this after finding the bugs.

**Comment by Justin900429** [comment]
The first thought of this would be the file name.
In my original setting, I sorted the files with their names which was five-digit base.
This was hard coding into the file and the filename should abide by this rule to ensure it works.

I will upload the rename file but need some time for it.

**Comment by tvaranka** [comment]
Here is an example of the output from the samm_crop (did my best to anonymize it :D). I think this is the issue. The file names shouldn't matter as they are getting loaded just fine.

Does the cropping look anything like what you have?
![image](https://user-images.githubusercontent.com/32382096/194475696-3f889b5f-4d7f-4db3-b8c4-2ca08f898227.png)

**Comment by Justin900429** [comment]
Could you print the `sorted_image_list` out for more info?
https://github.com/Justin900429/mimicking-annotation-micro-expression-recognition/blob/01fc04c168d849cd2eaa31943b184ae9f2788a9e/utils/samm_crop.py#L43

Want to check which image the landmark detection will work on

**Comment by tvaranka** [comment]
First two iterations:

['006_05562.jpg', '006_05563.jpg', '006_05564.jpg', '006_05565.jpg', '006_05566.jpg', '006_05567.jpg', '006_05568.jpg', '006_05569.jpg', '006_05570.jpg', '006_05571.jpg', '006_05572.jpg', '006_05573.jpg', '006_05574.jpg', '006_05575.jpg', '006_05576.jpg', '006_05577.jpg', '006_05578.jpg', '006_05579.jpg', '006_05580.jpg', '006_05581.jpg', '006_05582.jpg', '006_05583.jpg', '006_05584.jpg', '006_05585.jpg', '006_05586.jpg', '006_05587.jpg', '006_05588.jpg', '006_05589.jpg', '006_05590.jpg', '006_05591.jpg', '006_05592.jpg', '006_05593.jpg', '006_05594.jpg', '006_05595.jpg', '006_05596.jpg', '006_05597.jpg', '006_05598.jpg', '006_05599.jpg', '006_05600.jpg', '006_05601.jpg', '006_05602.jpg', '006_05603.jpg', '006_05604.jpg', '006_05605.jpg', '006_05606.jpg', '006_05607.jpg', '006_05608.jpg', '006_05609.jpg', '006_05610.jpg', '006_05611.jpg', '006_05612.jpg', '006_05613.jpg', '006_05614.jpg', '006_05615.jpg', '006_05616.jpg', '006_05617.jpg', '006_05618.jpg', '006_05619.jpg', '006_05620.jpg', '006_05621.jpg', '006_05622.jpg', '006_05623.jpg', '006_05624.jpg', '006_05625.jpg', '006_05626.jpg', '006_05627.jpg', '006_05628.jpg', '006_05629.jpg', '006_05630.jpg', '006_05631.jpg', '006_05632.jpg']

['006_03912.jpg', '006_03913.jpg', '006_03914.jpg', '006_03915.jpg', '006_03916.jpg', '006_03917.jpg', '006_03918.jpg', '006_03919.jpg', '006_03920.jpg', '006_03921.jpg', '006_03922.jpg', '006_03923.jpg', '006_03924.jpg', '006_03925.jpg', '006_03926.jpg', '006_03927.jpg', '006_03928.jpg', '006_03929.jpg', '006_03930.jpg', '006_03931.jpg', '006_03932.jpg', '006_03933.jpg', '006_03934.jpg', '006_03935.jpg', '006_03936.jpg', '006_03937.jpg', '006_03938.jpg', '006_03939.jpg', '006_03940.jpg', '006_03941.jpg', '006_03942.jpg', '006_03943.jpg', '006_03944.jpg', '006_03945.jpg', '006_03946.jpg', '006_03947.jpg', '006_03948.jpg', '006_03949.jpg', '006_03950.jpg', '006_03951.jpg', '006_03952.jpg', '006_03953.jpg', '006_03954.jpg', '006_03955.jpg', '006_03956.jpg', '006_03957.jpg', '006_03958.jpg', '006_03959.jpg', '006_03960.jpg', '006_03961.jpg', '006_03962.jpg', '006_03963.jpg', '006_03964.jpg', '006_03965.jpg', '006_03966.jpg', '006_03967.jpg', '006_03968.jpg', '006_03969.jpg', '006_03970.jpg', '006_03971.jpg', '006_03972.jpg', '006_03973.jpg', '006_03974.jpg', '006_03975.jpg', '006_03976.jpg', '006_03977.jpg', '006_03978.jpg', '006_03979.jpg', '006_03980.jpg', '006_03981.jpg', '006_03982.jpg', '006_03983.jpg', '006_03984.jpg', '006_03985.jpg', '006_03986.jpg', '006_03987.jpg', '006_03988.jpg']

**Comment by tvaranka** [comment]
Its cropping all the images, even the ones with four digits. I think the issue might be in the values used in the crop_region function. What size are the original images you have? Mine are 960x650
https://github.com/Justin900429/mimicking-annotation-micro-expression-recognition/blob/01fc04c168d849cd2eaa31943b184ae9f2788a9e/utils/samm_crop.py#L17-L21

**Comment by Justin900429** [comment]
I just find the code is not hard cording, it can work for any digit haha.
I'll retest this function and provide the correct version asap.

**Comment by Justin900429** [comment]
Hi, I had tried figurig out the problem but the file seems to work fine on my machine.

However, before doing our cropping method, the images are cropped out the face with [MTCNN](https://github.com/timesler/facenet-pytorch).
After doing this, the SAMM cropping might work well in your case.

I'll test whether this code is available for none-face-crop image. This might be the problem for the **dlib** landmark detection because their example requires an image includes only the face. Hence, they use the shape detector first. [Code link](https://github.com/davisking/dlib/blob/master/python_examples/face_landmark_detection.py#L86-L87)

```python
67  detector = dlib.get_frontal_face_detector()
...
86  # Get the landmarks/parts for the face in box d.
87  shape = predictor(img, d)
```
You can try this instead of using MTCNN, I guess this can also give the correct result.
(I updated the `samm_crop.py` and removed those constants and added `tqdm` but the functionality is still the same)

**Comment by tvaranka** [comment]
I tried the cropping with dlib and training myself. I was able to get 0.7854 F1-Score on SAMM three classes, which is quite close to the original 0.816. However, I noticed during the training that the test accuracy and f1 are changing rapidly in every epoch. In some cases even oscillating between 0 and 1. Did you face a similar issue, do you know what might cause it?

**Comment by tvaranka** [comment]
I think the reason for the rapidly changing f1 is here.
https://github.com/Justin900429/mimicking-annotation-micro-expression-recognition/blob/b0317b94142c1a27bd06c95c473127f42d4504f0/train.py#L88-L95

This is early stopping inside a K-Fold. If I run the training without early stopping I only get an F1-Score of 0.3452.

Have you experimented without early stopping?

**Comment by Justin900429** [comment]
The early stopping is not tested and this setting is based on all the above MER (capsule net, AU-ICGAN etc...)
The fastly changing is occurred due to the small number of valuation labels. MER model might change its result due to the model update and our future work will discuss this problem and propose some new methods to overcome this and achieve better accuracy with fewer update iterations.

**Comment by tvaranka** [comment]
I know many methods use it like you mentioned, however it is not correct. There is a data leak that gives a large positive bias to the result.

If you want to use early stopping yoy need to do it on seperate validation data, not the testing data

**Comment by Justin900429** [comment]
Other alternative could be tested on other MER dataset (cross dataset).
For example, SAMM (train) and validate (CASME). 
But this can also give a really incorrect result, because SAMM and CASME II can be in different manifold.

Currently, we have been exploring a new training approach and can overcome the oscillation and also provide even better early stopping.

**Comment by tvaranka** [comment]
Very interesting and hope to hear from those soon! Cross dataset is something I have also been working on lately. However, you can still have this faulty early stopping there so it doesn't fix the issue entirely.

**Comment by Justin900429** [comment]
Nope. Future work will not reuse the above-mentioned early stopping and will have a new setting.
